### PR TITLE
WIP CRM-20772 Allow higher precision decimals to be entered so tax can be calculate…

### PR DESCRIPTION
**Pull request by mattwire**
…d correctly

### CODE REVIEW

**Comment by mattwire** [comment]
Adding previous comments from @KarinG: I also see Invoice edits, address bits - and at the risk of sounding like @eileenmcnaughton I don't see a test that shows that shows the current problem (and that it is fixed with your PR); :-)

You can convince me of storing/carrying more decimals - absolutely - however I'm not following how the "value is copied from civicrm_price_field_value"? civicrm_membership_type.minimum_fee is set in the Membership Type config and for me is decimal(20,2).

Happy to review / help you think this through if you can:

strip this down to math only
add a test
make sure it does not fail on existing tests (as it does now);
I'd also be looking for a safe/minimum edits solution (first pass): i.e. the minimum code edits required to enable storing/carrying more decimals.

**Comment by mattwire** [comment]
@KarinG: This is pretty much the basics.  Leave it with me and I'll get a test added, clarify whether any of the other bits can be removed.

**Comment by mattwire** [comment]
@KarinG Ref the civicrm_membership_type.minimum_fee field type.  I discovered that the minimum_fee is actually stored as a price set and hence in civicrm_price_field_value (See: https://github.com/civicrm/civicrm-core/blob/276bd3ba20c20726c4ff535d143eeaa11c0f3b4a/CRM/Member/BAO/MembershipType.php#L711).  It is the civicrm_price_field_value.amount field which is specified as varchar(512) (See: https://github.com/civicrm/civicrm-core/blob/master/xml/schema/Price/PriceFieldValue.xml#L96).

Do you have any suggestions as to the best way to solve this?  The current type (Decimal 20,2) limits the minimum fee to 2 decimals so needs to change to allow greater precision.  I do agree that varchar(512) seems a little crazy!

File: xml/schema/Member/MembershipType.xml [hunk 1]
```diff
@@ -115,7 +115,8 @@
   <field>
     <name>minimum_fee</name>
     <title>membership Type Minimum Fee</title>
-    <type>decimal</type>
+    <type>varchar</type>
+    <length>512</length>
```
Comment (eileenmcnaughton) [inline xml/schema/Member/MembershipType.xml]: why change to varchar

File: CRM/Utils/Rule.php [hunk 2]
```diff
@@ -574,7 +574,10 @@ public static function money($value) {
       return TRUE;
     }
 
-    return preg_match('/(^-?\d+\.\d?\d?$)|(^-?\.\d\d?$)/', $value) ? TRUE : FALSE;
+    // Allow values such as -0, 1.024555, -.1
+    // We need to support multiple decimal places here, not just the number allowed by locale
+    //  otherwise tax calculations break when you want the inclusive amount to be a round number (eg. £10 inc. VAT requires 8.333333333 here).
+    return preg_match('/(^-?\d+\.?\d*$)|(^-?\.\d+$)/', $value) ? TRUE : FALSE;
```
Comment (eileenmcnaughton) [inline CRM/Utils/Rule.php]: I have a vested interest in this change ... https://issues.civicrm.org/jira/browse/CRM-17157

File: CRM/Upgrade/Incremental/Base.php [hunk 3]
```diff
@@ -182,6 +182,51 @@ public static function addColumn($ctx, $table, $column, $properties, $localizabl
...
+   *
+   * @param CRM_Queue_TaskContext $ctx
+   * @param string $table
+   * @param string $column
+   * @param string $properties
+   * @param bool $localizable is this a field that should be localized
+   * @param string $newColumn (new name for column)
+   * @return bool
+   */
+  public static function alterColumn($ctx, $table, $column, $properties, $localizable = FALSE, $newColumn = NULL) {
```
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/Base.php]: @seamuslee001 you are more on top of where we are at with these upgrade helpers - how does this fit in?

**Comment by eileenmcnaughton** [comment]
Can't we use DECIMAL(30,10)  or something like that?

https://dev.mysql.com/doc/refman/5.7/en/precision-math-decimal-characteristics.html

File: CRM/Upgrade/Incremental/Base.php [hunk 4]
```diff
@@ -182,6 +182,51 @@ public static function addColumn($ctx, $table, $column, $properties, $localizabl
...
+    if (!isset($newColumn)) {
+      // Option to rename column
+      $newColumn = $column;
+    }
+    $domain = new CRM_Core_DAO_Domain();
+    $domain->find(TRUE);
+    $queries = array();
+    if (!CRM_Core_BAO_SchemaHandler::checkIfFieldExists($table, $column)) {
```
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/Base.php]: I feel like here we should be checking that it exists

Comment (seamuslee001) [inline CRM/Upgrade/Incremental/Base.php]: Note that we may need new param on that function to say if we should be checking for translated name or not

File: CRM/Upgrade/Incremental/Base.php (hunk 3 above)
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/Base.php]: I think this makes sense to me i added in a couple of comments

File: CRM/Upgrade/Incremental/Base.php (hunk 4 above)
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/Base.php]: does $localizable do that?

Comment (seamuslee001) [inline CRM/Upgrade/Incremental/Base.php]: I'm not sure if that is the case on checkIfFieldExists, that is the one where we need to test are we looking for say help_pre_en_US or help_pre when we do that check

Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/Base.php]: Ah yes - so add localisable to that sig & pass it through ....

File: CRM/Utils/Rule.php (hunk 2 above)
Comment (mattwire) [inline CRM/Utils/Rule.php]: @eileenmcnaughton I could pull it out into it's own PR and we could try and get this in first?  It's crucial to allowing more than 2dp.

File: CRM/Upgrade/Incremental/Base.php (hunk 4 above)
Comment (mattwire) [inline CRM/Upgrade/Incremental/Base.php]: So, checkIfFieldExists does this: "SHOW COLUMNS FROM $tableName LIKE %1",
      array(1 => array($columnName, 'String')) which would catch localisable versions I think?
I pretty much copied AddColumn for this function, so it does whatever that helper does.

File: xml/schema/Member/MembershipType.xml (hunk 1 above)
Comment (mattwire) [inline xml/schema/Member/MembershipType.xml]: MembershipType.minimum_fee gets saved in civicrm_price_field_value.amount and retrieved from there.  That field is a varchar(512).  Which is why I used varchar(512), but anything that allows more decimals is fine by me :-)

File: CRM/Upgrade/Incremental/Base.php (hunk 4 above)
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/Base.php]: So the difference here is that unlike with addColumn we need to check that the original column exists in the db. In regards to the Show it looks like it may translate the column name properly but we should probably add in tests here https://github.com/civicrm/civicrm-core/blob/b8fdc7b55458438c9e60008b954b0f9243db0388/tests/phpunit/CRM/Core/BAO/SchemaHandlerTest.php#L150 to verify that it does the appropriate re-writing

File: CRM/Utils/Rule.php (hunk 2 above)
Comment (eileenmcnaughton) [inline CRM/Utils/Rule.php]: I feel like we could @JoeMurray @KarinG any thoughts? WMF made a change much like this a long time ago due to Cryptocurrencies & some middle eastern ones & we have been running with it. From what I can tell the DB field actually drops additional decimal places if it uses DECIMAL(10,2) - so this Rule change allows us to accept more decimal places, but it will only result in change if the DB field supports it

Comment (KarinG) [inline CRM/Utils/Rule.php]: Yes agreed - this is a good thing - no-one can argue with this.

File: xml/schema/Member/MembershipType.xml (hunk 1 above)
Comment (KarinG) [inline xml/schema/Member/MembershipType.xml]: Wrapping my head around the "MembershipType.minimum_fee gets saved in civicrm_price_field_value.amount and retrieved from there" - why... where..  is this triggered in the GUI?

Comment (eileenmcnaughton) [inline xml/schema/Member/MembershipType.xml]: When you edit membership min_amount it creates a default price set / updates one with a value for the membership type - so that goes there. Not sure why civicrm_price_field_value.amount  would be  varchar!

Comment (KarinG) [inline xml/schema/Member/MembershipType.xml]: Wicked!

**Comment by seamuslee001** [comment]
Jenkins ok to test this

File: CRM/Utils/Rule.php (hunk 2 above)
Comment (mattwire) [inline CRM/Utils/Rule.php]: See PR https://github.com/civicrm/civicrm-core/pull/10827

File: xml/schema/Member/MembershipType.xml (hunk 1 above)
Comment (mattwire) [inline xml/schema/Member/MembershipType.xml]: Posted this to dev list:
There are also currencies with more than two decimal places (https://en.wikipedia.org/wiki/ISO_4217#Active_codes) and bitcoin supports even more.

Now, the problem is that most amounts are stored in the DB as decimal(10,2) which means any precision greater than 2 decimal places is lost.  For this particular issue (CRM-20772) I am only proposing changing a single field in MembershipType.minimum_fee as that is the only one that impacts on this issue (civicrm_price_field_value.amount is where most monetary values end up and is stored as varchar(512) (though probably shouldn't be)).

What I would like to identify, is what is a sensible datatype that I should change minimum_fee to, and that we should standardise on in future for monetary amounts?

1. Most fields holding monetary amounts in the db are decimal(10,2), some are decimal(20,2), tax_rate is decimal(10,8).
1. We should probably change civicrm_price_field_value.amount from varchar(512) to something more sensible - but that's probably too risky a change at this stage.
2. We should define a datatype that should be used to hold monetary amounts that supports more than the current two decimals.
3. The current datatype decimal(10,2) supports a maximum value of ten million: 10,000,000.00
4. Decimal(30,10) would support 10,000,000,000,000,000.0000000000
5. MySQL stores 9 decimal places in 4 bytes, decimals and fractions separately (https://dev.mysql.com/doc/refman/5.7/en/storage-requirements.html) meaning multiples of 9 are the most sensible values to use.
5. Currencies support up to 4 decimal places, bitcoin(BTC) apparently support 8 decimal places.  So following GAAP accounting principals 10 decimal places gives the maximum required precision.  So, 9 or 18 for decimal places?
6. For decimals, 9 gives 100,000,000 but we currently store 10, 18 gives 100,000,000,000,000,000 (one hundred thousand million million) which should be plenty?

So: decimal(18,9)?

File: CRM/Member/Form/MembershipConfig.php [hunk 5]
```diff
@@ -83,7 +83,8 @@ public function setDefaultValues() {
     }
 
     if (isset($defaults['minimum_fee'])) {
-      $defaults['minimum_fee'] = CRM_Utils_Money::format($defaults['minimum_fee'], NULL, '%a');
+      // Don't use CRM_Utils_Money::format here as we need to allow more decimal places so sales tax can be calculated with recurring amounts.
+      $defaults['minimum_fee'] = filter_var($defaults['minimum_fee'], FILTER_SANITIZE_NUMBER_FLOAT, FILTER_FLAG_ALLOW_FRACTION);
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/MembershipConfig.php]: Once we've resolved the data format we should train CRM_Utils_Money::format to accept more decimals when appropriate, possibly another parameter to the function, or else wrap it - ie CRM_Utils_Money::formatLongDecimal

**Comment by JoeMurray** [comment]
This is coming along nicely. Thanks for all the work @mattwire

**Comment by eileenmcnaughton** [comment]
Have you seen this? https://github.com/moneyphp/money/issues

It doesn't support php 7.x  - although it might be that it does but has not been tested on it / declared to work on it. But a lot of thought has gone into how to handle money there

**Comment by JoeMurray** [comment]
I had bumped into that libary some time ago, but it seem on first glance like it would be too big an effort to include. However, perhaps we should re-evaluate if we are serious about trying to improve support for localization of currencies particularly with respect to number of digits after the separator.

**Comment by eileenmcnaughton** [comment]
@mattwire I'm going to close this as it is stale & a duplicate of a more recent one - where I think we should look to make progress is on the Membership Type form https://github.com/civicrm/civicrm-core/pull/10641/files#diff-9dacd15500c58030b03a70b7501fa5e5R86 - which is this field

$defaults['minimum_fee'] = CRM_Utils_Money::formatLongDecimal($defaults['minimum_fee']);

- we should get the format from the metadata of the field 'somehow' and then test that it appears correctly for both types of thousand separators (& try to keep the scope small).

We DID get stung changing the storage of amount on that field & we should tread carefully - but at the same time I think if we tackle one usage of money formatting at a time and a very clear about the context we should be OK - since the issues we had were not around the input & output so much as the deeper handling
